Add cronjob for sending reminders to patrons that have not yet retrieved loan

taskula opened the pull request.

taskula commented:
I'm not exactly sure if this logic was correct for figuring out whether or not patron has checked out the material. Essentially the logic in above commit is "illrequests with status IN_ANK and no active issues found with given biblionumber and borrowernumber". Or is there directly a suitable illrequests.status for such case?

MagnusEnger commented:
> I'm not exactly sure if this logic was correct for figuring out whether or not patron has checked out the material. Essentially the logic in above commit is "illrequests with status IN_ANK and no active issues found with given biblionumber and borrowernumber". Or is there directly a suitable illrequests.status for such case?

Yes, I think you can simplify that a bit:

There is a cronjob called cron/circstatus2illstatus.pl. It should be run often (every 2-3 minutes). It will look at requests with a status of IN_ANK (meaning the physical item has arrived at the library and been received by a librarian), and then check the connected record for items, and check the items for issues/loans. If the item is on loan, the ill_requests.status will be updated to IN_UTL. 

So if your script is run once every night, just looking for requests with a status of IN_ANK should be enough.

MagnusEnger commented on docs/cron.pod:
```diff
@@ -34,6 +35,9 @@ This is where the interesting stuff happens:
  /usr/sbin/koha-shell -c "/usr/bin/perl $CIRCSTAT > $LOGDIR/ill-libris-$DATETIME-circstat.log 2>&1" <instance>
  
+ # Send reminder notifications to patrons that have not yet retrieved the material
+ /usr/sbin/koha-shell -c "/usr/bin/perl $REDMINDER > $LOGDIR/ill-libris-$DATETIME-reminder.log 2>&1" <instance>
```
"$REDMINDER" looks like a typo

MagnusEnger commented on cron/reminder_notice.pl:
```diff
@@ -0,0 +1,118 @@
+    my $attributes = Koha::Illrequestattributes->search({
+        illrequest_id => $received->illrequest_id
+    });
+    next unless $attributes;
```
Looks like $attributes is not used anywhere?

taskula replied:
Removed in cf7e9ee

taskula commented:
> I think the problem is this line:
> 
> ```
> $reminders = $reminders->{ $type };
> ```
> 
> which destroys the original `$reminders` for subsequent runs of the `foreach my $received` loop?

Thanks for noticing. Resolved by b6dc2c0

taskula commented:
I also added some print statements for better logging of the cronjob in 67a99a0 and while fixing the issues you pointed out, I noticed another bug that caused the last configured reminder entry to be skipped due to broken array size comparison. be23324

MagnusEnger commented:
There is still some logic here I don't understand. 
```
    # find previous reminders per message transport type
    my $mtts = Koha::Illrequestattributes->search({
        illrequest_id => $received->illrequest_id,
        type => { -like => 'notified_mtt_%' }
    });

    my $reminder_count = Koha::Illrequestattributes->find({
        illrequest_id => $received->illrequest_id,
        type => 'notified_reminder_count'
    });
    $reminder_count = $reminder_count ? $reminder_count->value : 0;
    my $sent_reminder = 0;
    foreach my $mtt ( $mtts->as_list ) {
```
Does this mean that messages will only be sent through the message_transport_types that were used when the ill was received? And no reminder messages will be sent if no message was sent when it was received?

taskula commented:
> Does this mean that messages will only be sent through the message_transport_types that were used when the ill was received? And no reminder messages will be sent if no message was sent when it was received?

Yes, that was the idea. Now that you mention it, it does sound a bit silly if the library does not send notifications during receival. Then the reminder will not be sent at all. Should we force the reminder to be sent in any case? But then, with what message transport type(s)?

I suppose the most "correct" solution would be to let the patron configure messaging preferences for "ILL_Reminder" but this would then require a change in Koha's side for messaging preferences.

MagnusEnger commented:
> Should we force the reminder to be sent in any case? 

I would say "yes". 

> But then, with what message transport type(s)? I suppose the most "correct" solution would be to let the patron configure messaging preferences for "ILL_Reminder" but this would then require a change in Koha's side for messaging preferences.

Yes, that would be the perfect solution. But for now, I think we can just check what messages are configured. For a given letter code, if there is both an sms and email version, send both. If only one is set up, use that. 

The spec says:

>A reminder should be sent out automatically if the patron has not retrieved the material. At what time the reminder is generated should be regulated by a new system parameter. Through which channel (e-mail, print, sms) the reminder is sent is *regulated by Kohas template manager* (cgi-bin/koha/tools/letter.pl).

So I think this should be OK. (These messages should not be optional, in other words.)

taskula commented:
> But for now, I think we can just check what messages are configured. For a given letter code, if there is both an sms and email version, send both. If only one is set up, use that.

Okay, I've now implemented this. I've also squashed all the commits together for easier review.

Also, I noticed that letter.code has a maximum length of 20 and it is possible to run into bugs if you have too long letter code, e.g. the letter number could be trimmed out. So I've shortened the suggested letter codes to ILL_ANK_L1 (lån)/ILL_ANK_K1 (kopia) etc. Also I moved it to letter.module='ill' instead of 'circulation' so if you're testing it with letter templates from previous test runs, please note these changes.

MagnusEnger commented:
Works for me now. Thanks for fixing!
